Logbook tests are not reliable

<!-- READ THIS FIRST:
  - If you need additional help with this template, please refer to https://www.home-assistant.io/help/reporting_issues/
  - Make sure you are running the latest version of Home Assistant before reporting an issue: https://github.com/home-assistant/core/releases
  - Do not report issues for integrations if you are using custom components or integrations.
  - Provide as many details as possible. Paste logs, configuration samples and code into the backticks.
  DO NOT DELETE ANY TEXT from this template! Otherwise, your issue may be closed without comment.
-->
## The problem
<!-- 
  Describe the issue you are experiencing here to communicate to the
  maintainers. Tell us what you were trying to do and what happened.
-->

The following logbook tests are flapping

```
2020-09-22T20:58:35.5810061Z FAILED tests/components/logbook/test_init.py::TestComponentLogbook::test_service_call_create_logbook_entry
2020-09-22T20:58:35.5810835Z FAILED tests/components/logbook/test_init.py::test_logbook_describe_event[pyloop]
2020-09-22T20:58:35.5811488Z FAILED tests/components/logbook/test_init.py::test_exclude_described_event[pyloop]
```

```
2020-09-22T20:58:35.2512666Z =================================== FAILURES ===================================
2020-09-22T20:58:35.2513194Z _________ TestComponentLogbook.test_service_call_create_logbook_entry __________
2020-09-22T20:58:35.2514400Z [gw1] linux -- Python 3.8.2 /__w/core/core/venv/bin/python
2020-09-22T20:58:35.2514665Z 
2020-09-22T20:58:35.2515410Z self = <tests.components.logbook.test_init.TestComponentLogbook testMethod=test_service_call_create_logbook_entry>
2020-09-22T20:58:35.2516113Z 
2020-09-22T20:58:35.2516433Z     def test_service_call_create_logbook_entry(self):
2020-09-22T20:58:35.2517058Z         """Test if service call create log book entry."""
2020-09-22T20:58:35.2517380Z         calls = []
2020-09-22T20:58:35.2517584Z     
2020-09-22T20:58:35.2517817Z         @ha.callback
2020-09-22T20:58:35.2518115Z         def event_listener(event):
2020-09-22T20:58:35.2518421Z             """Append on event."""
2020-09-22T20:58:35.2518728Z             calls.append(event)
2020-09-22T20:58:35.2518983Z     
2020-09-22T20:58:35.2519437Z         self.hass.bus.listen(logbook.EVENT_LOGBOOK_ENTRY, event_listener)
2020-09-22T20:58:35.2520034Z         self.hass.services.call(
2020-09-22T20:58:35.2520445Z             logbook.DOMAIN,
2020-09-22T20:58:35.2520723Z             "log",
2020-09-22T20:58:35.2520930Z             {
2020-09-22T20:58:35.2521203Z                 logbook.ATTR_NAME: "Alarm",
2020-09-22T20:58:35.2521604Z                 logbook.ATTR_MESSAGE: "is triggered",
2020-09-22T20:58:35.2522013Z                 logbook.ATTR_DOMAIN: "switch",
2020-09-22T20:58:35.2522454Z                 logbook.ATTR_ENTITY_ID: "switch.test_switch",
2020-09-22T20:58:35.2522795Z             },
2020-09-22T20:58:35.2523006Z             True,
2020-09-22T20:58:35.2523213Z         )
2020-09-22T20:58:35.2523772Z         self.hass.services.call(
2020-09-22T20:58:35.2524183Z             logbook.DOMAIN,
2020-09-22T20:58:35.2524458Z             "log",
2020-09-22T20:58:35.2524664Z             {
2020-09-22T20:58:35.2524959Z                 logbook.ATTR_NAME: "This entry",
2020-09-22T20:58:35.2525386Z                 logbook.ATTR_MESSAGE: "has no domain or entity_id",
2020-09-22T20:58:35.2525721Z             },
2020-09-22T20:58:35.2525929Z             True,
2020-09-22T20:58:35.2526240Z         )
2020-09-22T20:58:35.2526575Z         # Logbook entry service call results in firing an event.
2020-09-22T20:58:35.2527069Z         # Our service call will unblock when the event listeners have been
2020-09-22T20:58:35.2527593Z         # scheduled. This means that they may not have been processed yet.
2020-09-22T20:58:35.2528049Z         self.hass.block_till_done()
2020-09-22T20:58:35.2528528Z         self.hass.data[recorder.DATA_INSTANCE].block_till_done()
2020-09-22T20:58:35.2528908Z     
2020-09-22T20:58:35.2529132Z         events = list(
2020-09-22T20:58:35.2529414Z             logbook._get_events(
2020-09-22T20:58:35.2529703Z                 self.hass,
2020-09-22T20:58:35.2530241Z                 dt_util.utcnow() - timedelta(hours=1),
2020-09-22T20:58:35.2530644Z                 dt_util.utcnow() + timedelta(hours=1),
2020-09-22T20:58:35.2530941Z             )
2020-09-22T20:58:35.2531122Z         )
2020-09-22T20:58:35.2531359Z >       assert len(events) == 2
2020-09-22T20:58:35.2531691Z E       AssertionError: assert 0 == 2
2020-09-22T20:58:35.2532004Z E        +  where 0 = len([])
2020-09-22T20:58:35.2532180Z 
2020-09-22T20:58:35.2532551Z tests/components/logbook/test_init.py:107: AssertionError
```

Fixed 
```
2020-09-22T20:58:35.5810835Z FAILED tests/components/logbook/test_init.py::test_logbook_describe_event[pyloop]
2020-09-22T20:58:35.5811488Z FAILED tests/components/logbook/test_init.py::test_exclude_described_event[pyloop]
``` in https://github.com/home-assistant/core/pull/40387

I'll fix

```
2020-09-22T20:58:35.5810061Z FAILED tests/components/logbook/test_init.py::TestComponentLogbook::test_service_call_create_logbook_entry
```


after https://github.com/home-assistant/core/pull/40070

--- a/homeassistant/components/recorder/__init__.py
+++ b/homeassistant/components/recorder/__init__.py
@@ -239,7 +239,7 @@ class Recorder(threading.Thread):
 
         self._timechanges_seen = 0
         self._keepalive_count = 0
-        self._old_state_ids = {}
+        self._old_states = {}
         self.event_session = None
         self.get_session = None
         self._completed_database_setup = False
@@ -385,7 +385,6 @@ class Recorder(threading.Thread):
                 if event.event_type == EVENT_STATE_CHANGED:
                     dbevent.event_data = "{}"
                 self.event_session.add(dbevent)
-                self.event_session.flush()
             except (TypeError, ValueError):
                 _LOGGER.warning("Event is not JSON serializable: %s", event)
             except Exception as err:  # pylint: disable=broad-except
@@ -396,16 +395,14 @@ class Recorder(threading.Thread):
                 try:
                     dbstate = States.from_event(event)
                     has_new_state = event.data.get("new_state")
-                    dbstate.old_state_id = self._old_state_ids.get(dbstate.entity_id)
+                    if dbstate.entity_id in self._old_states:
+                        dbstate.old_state = self._old_states.pop(dbstate.entity_id)
                     if not has_new_state:
                         dbstate.state = None
-                    dbstate.event_id = dbevent.event_id
+                    dbstate.event = dbevent
                     self.event_session.add(dbstate)
-                    self.event_session.flush()
                     if has_new_state:
-                        self._old_state_ids[dbstate.entity_id] = dbstate.state_id
-                    elif dbstate.entity_id in self._old_state_ids:
-                        del self._old_state_ids[dbstate.entity_id]
+                        self._old_states[dbstate.entity_id] = dbstate
                 except (TypeError, ValueError):
                     _LOGGER.warning(
                         "State is not JSON serializable: %s",

--- a/homeassistant/components/recorder/models.py
+++ b/homeassistant/components/recorder/models.py
@@ -14,6 +14,7 @@ from sqlalchemy import (
     distinct,
 )
 from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import relationship
 from sqlalchemy.orm.session import Session
 
 from homeassistant.core import Context, Event, EventOrigin, State, split_entity_id
@@ -105,7 +106,9 @@ class States(Base):  # type: ignore
     last_changed = Column(DateTime(timezone=True), default=dt_util.utcnow)
     last_updated = Column(DateTime(timezone=True), default=dt_util.utcnow, index=True)
     created = Column(DateTime(timezone=True), default=dt_util.utcnow)
-    old_state_id = Column(Integer)
+    old_state_id = Column(Integer, ForeignKey("states.state_id"))
+    event = relationship("Events", uselist=False)
+    old_state = relationship("States", remote_side=[state_id])
 
     __table_args__ = (
         # Used for fetching the state of entities at a specific time

--- a/tests/components/logbook/test_init.py
+++ b/tests/components/logbook/test_init.py
@@ -92,6 +92,7 @@ class TestComponentLogbook(unittest.TestCase):
         # Logbook entry service call results in firing an event.
         # Our service call will unblock when the event listeners have been
         # scheduled. This means that they may not have been processed yet.
+        trigger_db_commit(self.hass)
         self.hass.block_till_done()
         self.hass.data[recorder.DATA_INSTANCE].block_till_done()
 
@@ -1051,6 +1052,8 @@ async def test_logbook_describe_event(hass, hass_client):
     ):
         hass.bus.async_fire("some_event")
         await hass.async_block_till_done()
+        await hass.async_add_job(trigger_db_commit, hass)
+        await hass.async_block_till_done()
         await hass.async_add_executor_job(
             hass.data[recorder.DATA_INSTANCE].block_till_done
         )
@@ -1118,6 +1121,8 @@ async def test_exclude_described_event(hass, hass_client):
         hass.bus.async_fire(
             "some_event", {logbook.ATTR_NAME: name, logbook.ATTR_ENTITY_ID: entity_id3}
         )
+        await hass.async_block_till_done()
+        await hass.async_add_job(trigger_db_commit, hass)
         await hass.async_block_till_done()
         await hass.async_add_executor_job(
             hass.data[recorder.DATA_INSTANCE].block_till_done

--- a/tests/components/logbook/test_init.py
+++ b/tests/components/logbook/test_init.py
@@ -936,7 +936,7 @@ async def test_logbook_view(hass, hass_client):
     """Test the logbook view."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
     client = await hass_client()
     response = await client.get(f"/api/logbook/{dt_util.utcnow().isoformat()}")
     assert response.status == 200
@@ -946,7 +946,7 @@ async def test_logbook_view_period_entity(hass, hass_client):
     """Test the logbook view with period and entity."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     entity_id_test = "switch.test"
     hass.states.async_set(entity_id_test, STATE_OFF)
@@ -954,9 +954,9 @@ async def test_logbook_view_period_entity(hass, hass_client):
     entity_id_second = "switch.second"
     hass.states.async_set(entity_id_second, STATE_OFF)
     hass.states.async_set(entity_id_second, STATE_ON)
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1052,7 +1052,7 @@ async def test_logbook_describe_event(hass, hass_client):
     ):
         hass.bus.async_fire("some_event")
         await hass.async_block_till_done()
-        await hass.async_add_job(trigger_db_commit, hass)
+        await hass.async_add_executor_job(trigger_db_commit, hass)
         await hass.async_block_till_done()
         await hass.async_add_executor_job(
             hass.data[recorder.DATA_INSTANCE].block_till_done
@@ -1122,7 +1122,7 @@ async def test_exclude_described_event(hass, hass_client):
             "some_event", {logbook.ATTR_NAME: name, logbook.ATTR_ENTITY_ID: entity_id3}
         )
         await hass.async_block_till_done()
-        await hass.async_add_job(trigger_db_commit, hass)
+        await hass.async_add_executor_job(trigger_db_commit, hass)
         await hass.async_block_till_done()
         await hass.async_add_executor_job(
             hass.data[recorder.DATA_INSTANCE].block_till_done
@@ -1143,7 +1143,7 @@ async def test_logbook_view_end_time_entity(hass, hass_client):
     """Test the logbook view with end_time and entity."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     entity_id_test = "switch.test"
     hass.states.async_set(entity_id_test, STATE_OFF)
@@ -1151,9 +1151,9 @@ async def test_logbook_view_end_time_entity(hass, hass_client):
     entity_id_second = "switch.second"
     hass.states.async_set(entity_id_second, STATE_OFF)
     hass.states.async_set(entity_id_second, STATE_ON)
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1204,7 +1204,7 @@ async def test_logbook_entity_filter_with_automations(hass, hass_client):
     await async_setup_component(hass, "automation", {})
     await async_setup_component(hass, "script", {})
 
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     entity_id_test = "alarm_control_panel.area_001"
     hass.states.async_set(entity_id_test, STATE_OFF)
@@ -1223,9 +1223,9 @@ async def test_logbook_entity_filter_with_automations(hass, hass_client):
     )
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
 
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1276,7 +1276,7 @@ async def test_filter_continuous_sensor_values(hass, hass_client):
     """Test remove continuous sensor events from logbook."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     entity_id_test = "switch.test"
     hass.states.async_set(entity_id_test, STATE_OFF)
@@ -1288,9 +1288,9 @@ async def test_filter_continuous_sensor_values(hass, hass_client):
     hass.states.async_set(entity_id_third, STATE_OFF, {"unit_of_measurement": "foo"})
     hass.states.async_set(entity_id_third, STATE_ON, {"unit_of_measurement": "foo"})
 
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1312,7 +1312,7 @@ async def test_exclude_new_entities(hass, hass_client):
     """Test if events are excluded on first update."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     entity_id = "climate.bla"
     entity_id2 = "climate.blu"
@@ -1322,9 +1322,9 @@ async def test_exclude_new_entities(hass, hass_client):
     hass.states.async_set(entity_id2, STATE_OFF)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
 
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1347,7 +1347,7 @@ async def test_exclude_removed_entities(hass, hass_client):
     """Test if events are excluded on last update."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     entity_id = "climate.bla"
     entity_id2 = "climate.blu"
@@ -1363,9 +1363,9 @@ async def test_exclude_removed_entities(hass, hass_client):
     hass.states.async_remove(entity_id)
     hass.states.async_remove(entity_id2)
 
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1389,7 +1389,7 @@ async def test_exclude_attribute_changes(hass, hass_client):
     """Test if events of attribute changes are filtered."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
 
@@ -1402,9 +1402,9 @@ async def test_exclude_attribute_changes(hass, hass_client):
 
     await hass.async_block_till_done()
 
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1432,7 +1432,7 @@ async def test_logbook_entity_context_id(hass, hass_client):
     await async_setup_component(hass, "automation", {})
     await async_setup_component(hass, "script", {})
 
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     context = ha.Context(
         id="ac5bd62de45711eaaeb351041eec8dd9",
@@ -1472,7 +1472,7 @@ async def test_logbook_entity_context_id(hass, hass_client):
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     await hass.async_block_till_done()
 
-    await hass.async_add_job(
+    await hass.async_add_executor_job(
         logbook.log_entry,
         hass,
         "mock_name",
@@ -1483,7 +1483,7 @@ async def test_logbook_entity_context_id(hass, hass_client):
     )
     await hass.async_block_till_done()
 
-    await hass.async_add_job(
+    await hass.async_add_executor_job(
         logbook.log_entry,
         hass,
         "mock_name",
@@ -1518,9 +1518,9 @@ async def test_logbook_entity_context_id(hass, hass_client):
     )
     await hass.async_block_till_done()
 
-    await hass.async_add_job(trigger_db_commit, hass)
+    await hass.async_add_executor_job(trigger_db_commit, hass)
     await hass.async_block_till_done()
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     client = await hass_client()
 
@@ -1851,7 +1851,7 @@ async def test_icon_and_state(hass, hass_client):
     """Test to ensure state and custom icons are returned."""
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", {})
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
 
@@ -1900,7 +1900,7 @@ async def test_exclude_events_domain(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -1940,7 +1940,7 @@ async def test_exclude_events_domain_glob(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -1980,7 +1980,7 @@ async def test_include_events_entity(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -2013,7 +2013,7 @@ async def test_exclude_events_entity(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -2047,7 +2047,7 @@ async def test_include_events_domain(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -2091,7 +2091,7 @@ async def test_include_events_domain_glob(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -2143,7 +2143,7 @@ async def test_include_exclude_events(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)
@@ -2197,7 +2197,7 @@ async def test_include_exclude_events_with_glob_filters(hass, hass_client):
     )
     await hass.async_add_executor_job(init_recorder_component, hass)
     await async_setup_component(hass, "logbook", config)
-    await hass.async_add_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
+    await hass.async_add_executor_job(hass.data[recorder.DATA_INSTANCE].block_till_done)
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_START)
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STARTED)

--- a/tests/components/recorder/test_init.py
+++ b/tests/components/recorder/test_init.py
@@ -181,7 +181,20 @@ def test_saving_state_incl_entities(hass_recorder):
 
 def test_saving_event_exclude_event_type(hass_recorder):
     """Test saving and restoring an event."""
-    hass = hass_recorder({"exclude": {"event_types": "test"}})
+    hass = hass_recorder(
+        {
+            "exclude": {
+                "event_types": [
+                    "service_registered",
+                    "homeassistant_start",
+                    "component_loaded",
+                    "core_config_updated",
+                    "homeassistant_started",
+                    "test",
+                ]
+            }
+        }
+    )
     events = _add_events(hass, ["test", "test2"])
     assert len(events) == 1
     assert events[0].event_type == "test2"

